File: py/structure/tree/binary_tree/traversal/morris/PostOrderTraversal.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class TreeNode:
    def __init__(self, value: int):
        self.value, self.left, self.right = value, None, None

class PostOrderTraversal:
    def traversal(self, root: TreeNode) -> List[int]:
        result, current = [], root
        while current:
            if not current.left:
                current = current.right
                continue

            aux = current.left
            while aux.right and aux.right != current:
                aux = aux.right
            if not aux.right:
                aux.right, current = current, current.left
            else:
                aux.right = None
                self.__traversal(current.left, result)
                current = current.right
        self.__traversal(root, result)
        return result

    def __traversal(self, parent: TreeNode, result: List[int]) -> None:
        tail = self.__reverse(parent)
        current = tail
        while current:
            result.append(current.value)
            current = current.right
        self.__reverse(tail)

    def __reverse(self, current: TreeNode) -> TreeNode:
        prev = None
        while current:
            next = current.right
            current.right, prev, current = prev, current, next
        return prev

File: c/structure/tree/binary_tree/traversal/PostOrderTraversal.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct node {
    int value;
    struct node *left;
    struct node *right;
} TreeNode;

typedef TreeNode *pNode;

typedef struct {
    int *values;
    int length;
    int capacity;
} ArrayNode;

typedef struct stack {
    pNode value;
    struct stack *next;
} StackNode;

static StackNode *createStack();
static void pushStack(StackNode *stack, pNode value);
static pNode popStack(StackNode *stack);

void postOrderTraversal(pNode root, ArrayNode *result) {
    result->values = (int *) malloc(sizeof(int) * 64);
    memset(result->values, 0, sizeof(int) * 64);
    result->length = 0;
    result->capacity = 64;

    StackNode *stack = createStack();
    pushStack(stack, root);
    pNode current = NULL;
    while (stack->next) {
        pNode aux = stack->next->value;
        if (aux->left && current != aux->left && current != aux->right) {
            pushStack(stack, aux->left);
        } else if (aux->right && current != aux->right) {
            pushStack(stack, aux->right);
        } else {
            if (result->length >= result->capacity) {
                result->values = (int *) realloc(result->values, sizeof(int) * (result->capacity + 16));
                memset(result->values + result->capacity, 0, sizeof(int) * 16);
                result->capacity += 16;
            }
            result->values[result->length] = aux->value;
            ++result->length;

            current = aux;
            popStack(stack);
        }
    }
}

static StackNode *createStack() {
    StackNode *stack = (StackNode *) malloc(sizeof(StackNode));
    stack->value = NULL;
    stack->next = NULL;
    return stack;
}

static void pushStack(StackNode *stack, pNode value) {
    StackNode *node = (StackNode *) malloc(sizeof(StackNode));
    node->value = value;
    node->next = stack->next;
    stack->next = node;
}

static pNode popStack(StackNode *stack) {
    StackNode *node = stack->next;
    stack->next = node->next;
    return node->value;
}
